src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
-                        const UChar* lpTarget,
-                        int32_t cwTargetLength,
-                        const UChar* lpSource,
-                        int32_t cwSourceLength,
-                        int32_t options)
+ collation element is an int used for sorting. It consists of 3 components:
+    * primary - first 16 bits, representing the base letter
+    * secondary - next 8 bits, typically an accent
+    * tertiary - last 8 bits, typically the case
+
+An example (the numbers are made up to keep it simple)
+    a: 1 0 0
+    ą: 1 1 0
+    A: 1 0 1
+    Ą: 1 1 1
+
+    this method returns a mask that allows for characters comparison using specified Collator Strength
+*/
+static int32_t GetCollationElementMask(UColAttributeValue strength)
 {
-    int32_t result = FALSE;
-    UErrorCode err = U_ZERO_ERROR;
-    const UCollator* pColl = GetCollatorFromSortHandle(pSortHandle, options, &err);
+    switch (strength)
+    {
+        case UCOL_PRIMARY:
+            return UCOL_PRIMARYORDERMASK;
+        case UCOL_SECONDARY:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK;
+        default:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK | UCOL_TERTIARYORDERMASK;
+    }
+}
 
-    if (U_SUCCESS(err))
+static int32_t inline GetNextNonIgnorableCharacter(UCollationElements* pIterator, UErrorCode* pErrorCode)
+{
+    int32_t result = UCOL_IGNORABLE;
+
+    do
     {
-        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, &err);
-        int32_t idx = USEARCH_DONE;
+        result = ucol_next(pIterator, pErrorCode);
+    }
+    while (result == UCOL_IGNORABLE); // we don't check errorCode because on error the result is set to UCOL_NULLORDER
 
-        if (U_SUCCESS(err))
+    return result;
+}
+
+static int32_t SimpleStartsWith_Iterators(UCollationElements* pPatternIterator, UCollationElements* pSourceIterator, UColAttributeValue strength)
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+
+    int32_t collationElementMask = GetCollationElementMask(strength);
+
+    while (TRUE)
+    {
+        int32_t patternItem = GetNextNonIgnorableCharacter(pPatternIterator, &errorCode);
+        int32_t sourceItem = GetNextNonIgnorableCharacter(pSourceIterator, &errorCode);
+
+        if (patternItem == UCOL_NULLORDER)
         {
-            idx = usearch_first(pSearch, &err);
-            if (idx != USEARCH_DONE)
+            if (strength >= UCOL_SECONDARY && sourceItem != UCOL_NULLORDER)
             {
-                if (idx == 0)
+                if (((sourceItem & UCOL_PRIMARYORDERMASK) == 0) && (sourceItem & UCOL_SECONDARYORDERMASK) != 0)
                 {
-                    result = TRUE;
-                }
-                else
-                {
-                    result = CanIgnoreAllCollationElements(pColl, lpSource, idx);
+                    return FALSE; // the next non-ignorable character in source text is a combining character
```

Comment:
>return FALSE; // the next non-ignorable character in source text is a combining character [](start = 20, length = 89)

Do you have example of this case?

Reply:
Yes, `CultureInfo.InvariantCulture.CompareInfo.IsPrefix("o\u0308", "o")` provided by @GrabYourPitchforks